File: src/main/java/de/pflugradts/passbird/domain/service/password/RenamePasswordService.java
package de.pflugradts.passbird.domain.service.password;

import com.google.inject.Inject;
import de.pflugradts.passbird.domain.model.PasswordEntryNotFound;
import de.pflugradts.passbird.domain.model.password.KeyAlreadyExistsException;
import de.pflugradts.passbird.domain.model.transfer.Bytes;
import de.pflugradts.passbird.domain.service.eventhandling.EventRegistry;
import de.pflugradts.passbird.domain.service.password.encryption.CryptoProvider;
import de.pflugradts.passbird.domain.service.password.storage.PasswordEntryRepository;

import static de.pflugradts.passbird.domain.service.password.PasswordService.EntryNotExistsAction.CREATE_ENTRY_NOT_EXISTS_EVENT;

public class RenamePasswordService implements CommonPasswordServiceCapabilities {

    @Inject
    private CryptoProvider cryptoProvider;
    @Inject
    private PasswordEntryRepository passwordEntryRepository;
    @Inject
    private EventRegistry eventRegistry;

    public void renamePasswordEntry(final Bytes keyBytes, final Bytes newKeyBytes) {
        if (entryExists(cryptoProvider, passwordEntryRepository, eventRegistry, keyBytes, CREATE_ENTRY_NOT_EXISTS_EVENT)) {
            var encryptedNewKeyBytes = encrypted(cryptoProvider, newKeyBytes);
            if (find(passwordEntryRepository, encryptedNewKeyBytes).isEmpty()) {
                renamePasswordEntryOrFail(keyBytes, encryptedNewKeyBytes);
                processEventsAndSync(eventRegistry, passwordEntryRepository);
            } else {
                throw new KeyAlreadyExistsException(newKeyBytes);
            }
        }
    }

    private void renamePasswordEntryOrFail(final Bytes keyBytes, final Bytes newKeyBytes) {
        var encryptedKeyBytes = encrypted(cryptoProvider, keyBytes);
        find(passwordEntryRepository, encryptedKeyBytes).ifPresentOrElse(
                passwordEntry -> passwordEntry.rename(newKeyBytes),
                () -> eventRegistry.register(new PasswordEntryNotFound(encryptedKeyBytes)));
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/service/password/RenamePasswordService.kt
package de.pflugradts.passbird.domain.service.password
import de.pflugradts.kotlinextensions.TryResult
import de.pflugradts.kotlinextensions.TryResult.Companion.success
import de.pflugradts.passbird.domain.model.egg.EggIdAlreadyExistsException
import de.pflugradts.passbird.domain.model.shell.Shell
import de.pflugradts.passbird.domain.service.eventhandling.EventRegistry
import de.pflugradts.passbird.domain.service.password.PasswordService.EggNotExistsAction.CREATE_ENTRY_NOT_EXISTS_EVENT
import de.pflugradts.passbird.domain.service.password.encryption.CryptoProvider
import de.pflugradts.passbird.domain.service.password.tree.EggRepository
class RenamePasswordService constructor(
    cryptoProvider: CryptoProvider,
    eggRepository: EggRepository,
    eventRegistry: EventRegistry,
    memoryUpdateControl: MemoryUpdateControl,
) : CommonPasswordServiceCapabilities(cryptoProvider, eggRepository, eventRegistry, memoryUpdateControl) {
    fun renameEgg(eggIdShell: Shell, newEggIdShell: Shell): TryResult<Unit> {
        challengeEggId(newEggIdShell)
        if (eggExists(eggIdShell, CREATE_ENTRY_NOT_EXISTS_EVENT)) {
            if (find(newEggIdShell).isEmpty) {
                val egg = findWithoutUpdatingMemory(eggIdShell).get()
                val oldEggId = egg.viewEggId().copy()
                encrypted(newEggIdShell).let {
                    eggRepository.renameFavorites(egg.associatedNest(), oldEggId, it)
                    egg.rename(it)
                }
                return processEventsAndSync()
            } else {
                throw EggIdAlreadyExistsException(newEggIdShell)
            }
        }
        return success(Unit)
    }
}
